Fix Raspi GPIO binary_sensor produces unreliable responses

greg2001 opened the pull request:
Home Assistant's rpi_gpio binary sensor captures the pin state directly in the first egde interrupt.
At first glance, it looks logical and, on a real-time platform like Arduino, it would work properly.
On RPi/Linux, however, the timing is unpredictable, so that it might (actually, it often does) happen
that the signal bounces back before the edge handler has been called, and the handler captures a wrong pin
state. To fix it, we just wait until the signal stabilizes itself ("bouncetime" parameter) and capture
the pin state after that.

NOTE: PRi.GPIO's "bouncetime" parameter doesn't debounce input but simply disables additional egde interrupts
for the specified period of time.

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #10498
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

greg2001 committed:
Fix for issue #10498 Raspi GPIO binary_sensor produces unreliable responses ("Doorbell Scenario")

Home Assistant's rpi_gpio binary sensor captures the pin state directly in the first egde interrupt.
At first glance, it looks logical and, on a real-time platform like Arduino, it would work properly.
On RPi/Linux, however, the timing is unpredictable, so that it might (actually, it often does) happen
that the signal bounces back before the edge handler has been called, and the handler captures a wrong pin
state. To fix it, we just wait until the signal stabilizes itself ("bouncetime" parameter) and capture
the pin state after that.

NOTE: PRi.GPIO's "bouncetime" parameter doesn't debounce input but simply disables additional egde interrupts
for the specified period of time.

homeassistant commented:
Hi @greg2001, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#31788).

Once you do that we will be able to review and accept this pull request.

Thanks!

frenck commented:
Thanks, stalebot, but this PR awaits an initial review from us.
Please keep open for now, thanks 👍

frenck commented:
Meanwhile, @greg2001, the builds fail, which need to be addressed in order for this PR to be reviewed/merged. Could you take a look?

greg2001 commented:
> Meanwhile, @greg2001, the builds fail, which need to be addressed in order for this PR to be reviewed/merged. Could you take a look?

@frenck I would, but unfortunately, I have absolutely no idea about your build system. How do I get a detailed build log instead of just "Bash exited with code '1'"?

frenck commented:
The logs have been removed since it was quite a while back. I'll retrigger them.

frenck commented:
/AzurePipelines run

frenck commented:
Results: <https://dev.azure.com/home-assistant/Core/_build/results?buildId=33456&view=logs&jobId=f36f3843-8e6d-51cb-0e3a-71707ab37859&j=2fa2c639-d05e-530d-034e-4fb413fcea4e&t=88bd70b9-15d8-56b4-acb9-94f6fcce0e9d>
And: <https://dev.azure.com/home-assistant/Core/_build/results?buildId=33456&view=logs&jobId=4e214fed-f310-53fa-0018-b38e7beb405b&j=4e214fed-f310-53fa-0018-b38e7beb405b&t=6344b833-21e0-56d1-3a95-8a0fba7ad97a>

greg2001 committed:
Fix for issue #10498 Raspi GPIO binary_sensor produces unreliable responses ("Doorbell Scenario")

Home Assistant's rpi_gpio binary sensor captures the pin state directly in the first egde interrupt.
At first glance, it looks logical and, on a real-time platform like Arduino, it would work properly.
On RPi/Linux, however, the timing is unpredictable, so that it might (actually, it often does) happen
that the signal bounces back before the edge handler has been called, and the handler captures a wrong pin
state. To fix it, we just wait until the signal stabilizes itself ("bouncetime" parameter) and capture
the pin state after that.

NOTE: PRi.GPIO's "bouncetime" parameter doesn't debounce input but simply disables additional egde interrupts
for the specified period of time.

greg2001 commented:
@frenck Thanks, got it! Should be OK now.

balloob commented on homeassistant/components/rpi_gpio/binary_sensor.py:
```diff
@@ -62,12 +63,16 @@ def __init__(self, name, port, pull_mode, bouncetime, invert_logic):
+            """Edge detection handler."""
+            threading.Timer(float(self._bouncetime) / 1000, read_gpio).start()
+
+        rpi_gpio.edge_detect(self._port, edge_detected, self._bouncetime)
```
How often is an edge detected ?

balloob commented on homeassistant/components/rpi_gpio/binary_sensor.py:
```diff
@@ -62,12 +63,16 @@ def __init__(self, name, port, pull_mode, bouncetime, invert_logic):
-        rpi_gpio.edge_detect(self._port, read_gpio, self._bouncetime)
+        def edge_detected(port):
+            """Edge detection handler."""
+            threading.Timer(float(self._bouncetime) / 1000, read_gpio).start()
```
Let's not use a thread but instead schedule it on the event loop.

```suggestion
		async def async_read_gpio(self):
		    """Read state from GPIO."""
		    await asyncio.sleep(float(self._bouncetime) / 1000)
		    self._state = await self.hass.async_add_executor_job(rpio_gpio.read_input, self._port)
		    self.async_write_ha_state()

        def edge_detected(port):
            """Edge detection handler."""
            self.hass.add_job(self.async_read_gpio)
```

Misiu commented:
Not stale

frenck commented:
@Misiu I'm sorry, can I ask why you wrote not stale?
You have not been involved in this PR, and as a matter of fact, it is stale.

Misiu commented:
@frenck I was observing this PR from the beginning and I saw that @balloob give some reviews.
The is no comment that this shouldn't get merged.
Please give the author some more time to address the comments.
My last PI3 is broken, I'm waiting for a new one and I want to get involved in fixing this.

frenck commented:
@Misiu The author did not respond in a month. That is why it is marked stale. You've just messed with that process, which isn't appreciated. Please don't do that again.

I'm sorry to hear you have problems, but the stale bot is here for a reason.

greg2001 commented:
@frenck @balloob 
Just saw it closed. No interest in fixing this old problem?

frenck commented:
@greg2001 Sure we are!

However, this PR has been stale for over a month, without response. Hence it was closed. Feel free to re-open a PR when ready to work on it again and address the comments raised!

👍

greg2001 commented:
@frenck Would you kindly re-open this PR or shall I create a new one?

TheFax commented:
I would like to remember that issue #10498 concerns the same problem and is still open.
I worked on it 25 days ago.
